keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+
+To sum up. Some applications solve the disruption problem by introducing validating admission
+webhooks. This has some drawbacks. The webhooks are not easily discoverable by cluster admins. And
+they can block evictions for other applications if they are misconfigured or misbehave. The
+eviction API is not intended to be extensible in this way. The webhook approach is therefore not
+recommended.
+
+Some drainers solve the node drain by depending on the kubectl logic, or by extending/rewriting it
+with additional rules and logic.
+
+As seen in the experience reports and GitHub issues, some admins solve their problems by simply
+ignoring PDBs which can cause unnecessary disruptions or data loss. Some solve this by playing
+with the application deployment, but have to understand that the application supports this.
+
+kubelet's Graceful Node Shutdown feature is a best-effort solution for unplanned shutdowns, but
+it is not sufficient to ensure application and data safety.
+
+### Goals
+- Introduce NodeMaintenance API.
+- Introduce a node maintenance controller that creates evacuations.
+- Deprecate kubectl drain in favor of NodeMaintenance.
+- Implement NodeMaintenanceAwareKubelet feature to implement a lifecycle for static pods during a
+  maintenance.
+- Implement NodeMaintenanceAwareDaemonSet feature to prevent the scheduling of DaemonSet pods on
+  nodes during a maintenance.
+
+### Non-Goals
+- Introduce a node maintenance period, nodeDrainTimeout (similar to [cluster-api](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane)
+  nodeDrainTimeout) or a TTL optional field as an upper bound on the duration of node maintenance.
+  Then the node maintenance would be garbage collected and the node made schedulable again.
+- Solve the node lifecycle management or automatic shutdown after the node drain is completed.
+  Implementation of this is better suited for other cluster components and actors who can use the
+  node maintenance as a building block to achieve their desired goals.
+- Synchronize all pod termination mechanisms (see #8 in the [Motivation](#motivation) section), so that they do
+  not terminate pods under NodeMaintenance/Evacuation.
+
+## Proposal
+
+Most of these issues stem from the lack of a standardized way of detecting a start of the node
+drain. This KEP proposes the introduction of a NodeMaintenance object that would signal an intent
+to gracefully remove pods from given nodes. The intent will be implemented by the newly proposed
+[Evacuation API KEP](https://github.com/kubernetes/enhancements/issues/4563), which ensures
+graceful pod removal or migration, an ability to measure the progress and a fallback to eviction if
+progress is lost. The NodeMaintenance implementation should also utilize existing node's
+`.spec.unschedulable` field, which prevents new pods from being scheduled on such a node.
+
+We will deprecate the `kubectl drain` as the main mechanism for draining nodes and drive the whole
```

Comment:
With my sig-cli hat, I will call out this wasn't discussed with that group, so before proceeding with that kind of steps you'll need their approval.

Reply:
This is mostly symbolic, as the NodeMaintenance API will essentially replace all of its functionality. The command itself does not have to be removed. A warning that there is a better way, might be enough.

Nevertheless, I plan to present this to sig-cli before committing to it.

Reply:
I'd imagined that we'd keep `kubectl drain` and that by default this subcommand would make a NodeMaintenance if the API is available, otherwise fall back to client-side draining.